core/src/main/java/com/linecorp/armeria/common/util/DefaultAsyncLoader.java

```diff
@@ -0,0 +1,264 @@
+            logger.warn("Unexpected exception from exceptionHandler.apply()", e);
+            future.completeExceptionally(cause);
+        }
+    }
+
+    private boolean isValid(@Nullable CacheEntry<T> cacheEntry) {
+        if (cacheEntry == null) {
+            return false;
+        }
+
+        if (expireAfterLoadNanos > 0) {
+            final long elapsed = System.nanoTime() - cacheEntry.cachedAt;
+            if (elapsed >= expireAfterLoadNanos) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("The cached value expired after {} ms. cache: {}",
+                                 TimeUnit.NANOSECONDS.toMillis(expireAfterLoadNanos), cacheEntry.value);
+                }
+                return false;
+            }
+        }
+
+        if (expireIf != null && expireIf.test(cacheEntry.value)) {
+            logger.debug("The cached value expired due to 'expireIf' condition. cache: {}", cacheEntry.value);
+            return false;
+        }
+
+        return true;
+    }
+
+    private static class CacheEntry<T> {
+
+        private final T value;
+        private final long cachedAt = System.nanoTime();
+
+        CacheEntry(T value) {
+            requireNonNull(value, "value");
+            this.value = value;
+        }
+    }
+
+    private static class RefreshingFuture<U> extends CompletableFuture<CacheEntry<U>> {
+
+        private static final RefreshingFuture<?> COMPLETED;
+
+        static {
+            COMPLETED = new RefreshingFuture<>(null);
+            COMPLETED.complete(null);
+        }
+
+        @SuppressWarnings("unchecked")
+        static <T> RefreshingFuture<T> completedFuture() {
+            return (RefreshingFuture<T>) COMPLETED;
+        }
+
+        @Nullable
+        private volatile CacheEntry<U> cacheEntry;
+        // True when refreshing is in progress, means not yet completed by new value or exception.
+        private volatile boolean refreshing;
+
+        RefreshingFuture(@Nullable CacheEntry<U> cacheEntry) {
+            // cacheEntry can be null if it expired or has not loaded yet.
+            this.cacheEntry = cacheEntry;
+            if (cacheEntry != null) {
+                complete(cacheEntry);
+            }
+            refreshing = true;
+        }
+
+        @Override
+        public boolean complete(CacheEntry<U> newVal) {
+            refreshing = false;
+            cacheEntry = newVal;
+            obtrudeValue(newVal);
+            return true;
+        }
+
+        @Override
+        public boolean completeExceptionally(Throwable ex) {
+            refreshing = false;
+            final CacheEntry<U> cacheEntry = this.cacheEntry;
+            if (cacheEntry != null) {
+                logger.warn("Failed to refresh a new value. cache: {}", cacheEntry.value, ex);
+            } else {
+                obtrudeException(ex);
+            }
+            return true;
+        }
+
+        @Override
+        public CacheEntry<U> join() {
```

Comment:
Question: Do we need to override this method? `cacheEntry` is set to the result of the completable future anyway?

Reply:
oh~ I found that we don't need to override this `join()` cause we call `obtrudeValue() or obtrudeException()` anyway inside of overrided `complete() and completeExceptionally()` 👍  thanks!